src/Dscp/Crypto/Encrypt.hs

```diff
@@ -156,31 +157,47 @@ prepareAEAD (PassPhrase pp) =
        aeadInit aeadMode cipher initIV
 
 -- | Encrypt given 'ByteArray' with AES.
-encrypt :: FromByteArray ba => PassPhrase -> ba -> Encrypted ba
-encrypt pp plaintext =
+encryptBA :: ByteArray ba => PassPhrase -> ba -> Encrypted ba
+encryptBA pp plaintext =
     let aead = prepareAEAD pp
-        -- TODO: Subject to optimization. For now using converion to bytestring
-        -- since @ba@ is not instance of 'ByteArray'.
-        plaintextBS = BA.convert plaintext
-        (authtag, ciphertextBS :: ByteString) =
-            aeadSimpleEncrypt aead authHeader plaintextBS authTagLength
-        ciphertext =
-            leftToPanicWith "encrypt: got malformed item" $
-            fromByteArray ciphertextBS
-    in  Encrypted authtag ciphertext
+    in uncurry Encrypted $
+       aeadSimpleEncrypt aead authHeader plaintext authTagLength
 
 -- | Decrypt given 'Encrypted' datatype or fail, if passphrase
 -- doesn't match.
-decrypt :: FromByteArray ba => PassPhrase -> Encrypted ba -> Either DecryptionError ba
-decrypt pp (Encrypted tag ciphertext) = do
+decryptBA :: ByteArray ba => PassPhrase -> Encrypted ba -> Either DecryptionError ba
+decryptBA pp (Encrypted tag ciphertext) =
     let aead = prepareAEAD pp
-        ciphertextBS = BA.convert ciphertext :: ByteString
-    plaintextBS <-
-          maybeToRight PassPhraseInvalid $
-          aeadSimpleDecrypt aead authHeader ciphertextBS tag
-    return $
-        leftToPanicWith "decrypt: got malformed item" $
-        fromByteArray plaintextBS
+    in maybeToRight PassPhraseInvalid $
+       aeadSimpleDecrypt aead authHeader ciphertext tag
+
+-- | 'FromByteArray' versions of 'encryptBA' and 'decryptBA'.
+-- Default implementation is correct if the following holds:
+--
+-- * If some @ba@ is valid, then any @ba'@ of the same length
+-- has to be also valid.
+class EncryptSafe ba where
```

Comment:
Frankly, I think that this class is really unnecessary. I struggle to think about some special datatypes, which don't have `ByteArray` instance and cannot have `FromByteArray` one, and still can (and should) be encrypted in some fancy optimized way which doesn't involve serialisation.

Reply:
Yes, but point of this typeclass is to restrict usage of `encrypt` and `decrypt`, since we use this `FromByteArray` stuff and thus they are actually claimed unsafe. I can rename them somehow to show that they work with instances of `ByteArray`/`FromByteArray` to avoid confusion.